아이디찾기 + 비밀번호 찾기 API

leeys9423 opened the pull request:
## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->
- 아이디 찾기 및 비밀번호 초기화 기능을 구현하여 사용자가 이메일이나 이름을 통해 아이디와 비밀번호를 찾을 수 있도록 하는 기능을 추가하였습니다. 또한, 비밀번호 초기화 과정에서 발생할 수 있는 예외들을 처리하고, 관련된 클래스명 변경 및 리팩토링을 수행했습니다.

## 👩‍💻 요구 사항과 구현 내용 <!-- 기능을 Commit 별로 잘개 쪼개고, Commit 별로 설명해주세요 -->
- feat: 비밀번호 초기화 기능 구현 (RB-154)
비밀번호 초기화 기능을 구현하여, 사용자가 이메일과 이름을 입력하면 임시 비밀번호를 생성하여 해당 이메일로 전송합니다.
POST /auth/reset-password API를 호출하고, 비밀번호 변경 후 이메일로 안내를 전송하는 로직을 추가했습니다.

- feat: 비밀번호 초기화 예외처리 (RB-155)
비밀번호 초기화 시 발생할 수 있는 예외들을 처리하기 위한 로직을 추가했습니다.
예외 처리 로직을 통해 사용자가 잘못된 정보를 입력한 경우 적절한 오류 메시지를 반환하고, 비밀번호 초기화가 정상적으로 처리되지 않도록 방지했습니다.

- refactor: 비밀번호 초기화 예외처리 (RB-155)
예외 처리 로직을 리팩토링하여 비밀번호 초기화와 관련된 오류 메시지 및 예외 상황을 좀 더 명확하게 처리할 수 있도록 개선했습니다.

- feat: 아이디 찾기 기능 구현
사용자가 이름과 전화번호를 입력하여 아이디를 찾을 수 있는 기능을 추가했습니다.
POST /auth/find-id API를 호출하여 아이디(이메일)를 찾아주는 기능을 구현하고, 결과를 화면에 표시합니다.
로딩 상태 및 오류 메시지를 관리하여 사용자 경험을 개선했습니다.

- refactor: PasswordResetRequest 클래스명을 ResetPasswordRequest로 변경
PasswordResetRequest 클래스의 이름을 ResetPasswordRequest로 변경하여 코드의 일관성을 유지하고, 클래스 명칭을 더 직관적으로 수정했습니다.

jhyngu committed:
feat: 비밀번호 초기화 예외처리 (RB-155)
- EmptyPasswordException 구현
- ErrorCode : EMPTY_PASSWORD 추가
- User : 테스트용 생성자, changePassword() 추가

jhyngu committed:
feat: 비밀번호 초기화 기능 구현 (RB-154)
- PasswordResetRequest 구현
- temp-password-email.html 작성
- UserController : POST, uri = "/auth/password/reset" 구현
- UserRepository : findByEmailAndName() 구현
- UserService : resetPassword(), createTempPasswordEmailContent() 구현

jhyngu committed:
refactor: 비밀번호 초기화 예외처리 (RB-155)
- ErrorCode : USER_NOT_FOUND 추가
- UserNotFoundException 구현

jhyngu committed:
feat: 아이디 찾기 기능 구현

- 아이디 찾기(POST /auth/find/id) API 구현
 - 사용자 이름과 전화번호를 입력받아 이메일 반환
- UserService에 아이디 찾기 로직 구현
 - 사용자 이름, 전화번호 유효성 검사
 - 사용자 정보가 없을 경우 예외(InvalidUserInfoException, UserIdNotFoundException)를 활용하여 에러 처리
- UserController에 엔드포인트 추가
 - 요청 : FindIdRequest 구현
 - 응답 : FindIdResponse 구현
- ErrorCode에 새로운 에러 코드 추가
 - INVALID_USER_INFO : 잘못된 사용자 이름 또는 전화번호 입력한 경우
 - USER_ID_NOT_FOUND : 입력된 정보와 일치하는 사용자가 없을 경우
- UserRepository에 사용자 조회 쿼리 메서드 추가
 - findByNameAndPhoneNumber(String name, String phoneNumber)

jhyngu committed:
refactor: PasswordResetRequest 클래스명을 ResetPasswordRequest로 변경

leeys9423 commented on backend/src/main/java/com/redbox/domain/user/entity/User.java:
```diff
@@ -56,4 +58,18 @@ public User(String email, String password, String name, LocalDate birth, String
     }
 
+    // 테스트용 생성자
+    public User(String email, String name, String password) {
```
이 생성자 사용안하는데 지워도 될거같아요

jhyngu replied:
지웠습니다~

leeys9423 commented on backend/src/main/java/com/redbox/domain/user/exception/UserIdNotFoundException.java:
```diff
@@ -0,0 +1,10 @@
+import com.redbox.global.exception.BusinessException;
+import com.redbox.global.exception.ErrorCode;
+
+public class UserIdNotFoundException extends BusinessException {
```
UserNotFoundException 에러클래스랑 성격이 많이 겹치는거 같은데 한개로 사용해보시는 방향으로 해보시는건 어떠세요?

jhyngu replied:
저도 이 부분 고민했었는데 그냥 통합하겠습니다~

leeys9423 commented on backend/src/main/java/com/redbox/domain/user/service/UserService.java:
```diff
@@ -86,4 +86,50 @@ public SignupResponse signup(SignupRequest request) {
+        emailSender.sendMail(request.getEmail(), subject, content);
+    }
+
+    private String createTempPasswordEmailContent(String tempPassword) {
```
createAuthCodeEmailContent 이 메서드와 똑같이 생겨서 그런데, 공통적인 메서드를 하나 따로 만들어서 관리하는게 좋아보이는데 어떻게 생각하세요~?

jhyngu replied:
하나의 메서드로 관리하게 수정하겠습니다.

leeys9423 commented on backend/src/main/java/com/redbox/domain/user/service/UserService.java:
```diff
@@ -86,4 +86,50 @@ public SignupResponse signup(SignupRequest request) {
+        String phoneNumber = request.getPhoneNumber();
+
+        // 사용자 이름과 전화번호가 올바르게 입력되었는지 확인
+        if (name == null || phoneNumber == null || name.trim().isEmpty() || phoneNumber.trim().isEmpty()) {
```
이 부분은 일단 FindIdRequest dto 에서 validation을 활용해보는게 어떨까요?
서비스로직에서 확실한 Null체크를 하는것도 좋지만 validation을 활용하면 전 단계에서 BAD_REQUEST에 대한 에러 핸들링을 미리 할 수 있으니까 코드의 중복을 막을 수 있을 것 같다는 생각을 해봅니다~

jhyngu replied:
넵 validation 사용하겠습니다~

leeys9423 commented on backend/src/main/java/com/redbox/domain/user/dto/FindIdRequest.java:
```diff
@@ -0,0 +1,13 @@
+@Getter
+@NoArgsConstructor
+@AllArgsConstructor
+public class FindIdRequest {
```
validation을 활용하면 BAD_REQUEST 에러를 바로 터트릴 수 있슴니다

jhyngu replied:
넵👍

leeys9423 commented on backend/src/main/java/com/redbox/domain/user/service/UserService.java:
```diff
@@ -86,4 +86,50 @@ public SignupResponse signup(SignupRequest request) {
+
+        // 사용자 이름과 전화번호가 올바르게 입력되었는지 확인
+        if (name == null || phoneNumber == null || name.trim().isEmpty() || phoneNumber.trim().isEmpty()) {
+            throw new InvalidUserInfoException(); // 잘못된 사용자 정보 처리
```
그리고 Validation과 exception의 경계를 확실하게 분리하는게 좋아보여요.
검증 실패와 비즈니스 Exception 성격 자체가 다르거든요.
- Validation 실패 (@NotNull, @NotBlank 등)
클라이언트가 잘못된 형식이나 누락된 데이터를 보냈을 때
Spring의 MethodArgumentNotValidException으로 자동 처리
400 Bad Request로 처리하면 충분
- 비즈니스 Exception
형식은 올바르지만 비즈니스 규칙을 위반할 때
예: 이미 가입된 이메일, 잔액 부족, 재고 부족 등
커스텀 Exception으로 분리해서 관리

jhyngu replied:
네엡!!

jhyngu committed:
refactor: UserIdNotFoundException과 UserNotFoundException 중복 제거

- UserIdNotFoundException의 역할을 UserNotFoundException으로 통합

jhyngu committed:
refactor: 사용되지 않는 User 클래스의 테스트용 생성자 제거

- 코드 정리를 통해 불필요한 생성자 삭제

jhyngu committed:
refactor: 이메일 콘텐츠 생성 메서드 리팩토링

- 인증 코드 이메일 생성 메서드를 일반화하여 템플릿 이름과 변수명을 매개변수로 받는 `createEmailContent` 메서드로 변경

jhyngu committed:
feat: FindIdRequest DTO에 유효성 검증 추가 및 UserService 수정

- FindIdRequest DTO에 @NotBlank 및 @Pattern을 사용하여 입력 검증 추가
- UserService의 findUserId 메서드에서 중복된 유효성 검증 제거
- 컨트롤러에서 @Valid를 사용하여 유효성 검증 자동 처리

jhyngu committed:
Merge branch 'devBE' into feature/find-account
